bridge/setu/processor/base.go

```diff
@@ -96,3 +103,123 @@ func (bp *BaseProcessor) String() string {
 func (bp *BaseProcessor) Stop() {
 	// override to stop any go-routines in individual processors
 }
+
+// isOldTx checks if the transaction already exists in the chain or not
+// It is a generic function, which is consumed in all processors
+func (bp *BaseProcessor) isOldTx(cliCtx cliContext.CLIContext, txHash string, logIndex uint64) (bool, error) {
+	queryParam := map[string]interface{}{
+		"txhash":   txHash,
+		"logindex": logIndex,
+	}
+
+	endpoint := helper.GetHeimdallServerEndpoint(util.StakingTxStatusURL)
+	url, err := util.CreateURLWithQuery(endpoint, queryParam)
+	if err != nil {
+		bp.Logger.Error("Error in creating url", "endpoint", endpoint, "error", err)
+		return false, err
+	}
+
+	res, err := helper.FetchFromAPI(bp.cliCtx, url)
+	if err != nil {
+		bp.Logger.Error("Error fetching tx status", "url", url, "error", err)
+		return false, err
+	}
+
+	var status bool
+	if err := json.Unmarshal(res.Result, &status); err != nil {
+		bp.Logger.Error("Error unmarshalling tx status received from Heimdall Server", "error", err)
+		return false, err
+	}
+
+	return status, nil
+}
+
+// checkTxAgainstMempool checks if the transaction is already in the mempool or not
+// It is consumed only for `clerk` processor
+func (bp *BaseProcessor) checkTxAgainstMempool(msg types.Msg) (bool, error) {
+	endpoint := helper.GetConfig().TendermintRPCUrl + util.TendermintUnconfirmedTxsURL
+	resp, err := http.Get(endpoint)
+	if err != nil || resp.StatusCode != http.StatusOK {
+		bp.Logger.Error("Error fetching mempool tx", "url", endpoint, "error", err)
+		return false, err
+	}
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		bp.Logger.Error("Error fetching mempool tx", "error", err)
+		return false, err
+	}
+
+	// a minimal response of the unconfirmed txs
+	var Response struct {
+		Result struct {
+			Total string   `json:"total"`
+			Txs   []string `json:"txs"`
+		} `json:"result"`
+	}
+
+	err = json.Unmarshal(body, &Response)
+	if err != nil {
+		bp.Logger.Error("Error unmarshalling response received from Heimdall Server", "error", err)
+		return false, err
+	}
+
+	// Iterate over txs present in the mempool
+	// We can verify if the message we're about to send is present by
+	// checking the type of transaction, the transaction hash and log index
+	// present in the data of transaction
+
+	status := false
+Loop:
+	for _, txn := range Response.Result.Txs {
```

Comment:
Do we have some data about how big the mempool usually are and how many incoming state syncs events can burst up to? Getting all Txs from mempool and check every sync event against them with retry could have non-trivial impact on the performance (I could be wrong if the numbers are small).

Reply:
Usually mempool size is < 10 txns and we get up to 10-12 state sync events in a block, Max we can process as of now is 25 per Heimdall block. Your point is correct if we already have a huge number of txns in mempool but this check is actually implemented to prevent that only. We will also take a look at how we can improve this further.

Reply:
One thought which popped up my mind while implementing was to keep a map in memory as a cache for mempool transactions in a very specific format, which we require for the check. We can find any duplicate transaction through it in O(1).

Reply:
Good idea about the memory cache. I am fine with current implementation. We can revisit this and optimize it if performance got hit in the future.

Reply:
Sure. Resolving this for now.